### PR TITLE
report: truncate long attribute values in HTML snippets

### PR DESCRIPTION
**Pull request by Beytoven**
Addresses #10717.

Before:
Screen Shot 2020-07-14 at 10 49 09 AM

After:
Screen Shot 2020-07-14 at 10 41 18 AM

### CODE REVIEW

File: lighthouse-core/lib/page-functions.js [hunk 1]
```diff
@@ -123,6 +123,14 @@ function getOuterHTMLSnippet(element, ignoreAttrs = []) {
     ignoreAttrs.forEach(attribute =>{
       clone.removeAttribute(attribute);
     });
+    for (const attributeName of clone.getAttributeNames()) {
+      let attributeValue = clone.getAttribute(attributeName);
+      if (attributeValue.length > 100) {
```
Comment (paulirish) [inline lighthouse-core/lib/page-functions.js:128]: i'd pull this 100 up to a const towards the top of the method

File: lighthouse-core/lib/page-functions.js [hunk 2]
```diff
@@ -123,6 +123,14 @@ function getOuterHTMLSnippet(element, ignoreAttrs = []) {
...
+        attributeValue = attributeValue.slice(0, 97) + '...';
```
Comment (paulirish) [inline lighthouse-core/lib/page-functions.js:129]: protip:  `option-;` on mac is `…` which is a single character ellipsis.

**Comment by brendankenny** [comment]
before/after screenshots? :)

**Comment by brendankenny** [comment]
> before/after screenshots? :)

oh, I see, not limiting which attributes, just making sure none of them are ridiculously long. I guess that's also not that hard to imagine :)

**Comment by connorjclark** [comment]
This is technically a breaking change.

Any way we could truncate in just the report but leave `snippet` untouched?

**Comment by Beytoven** [comment]
> This is technically a breaking change.
> 
> Any way we could truncate in just the report but leave `snippet` untouched?

I could look into making this change in the snippet renderer instead. I think eventually we'd want to do it in page-functions for the next major release.

**Comment by connorjclark** [comment]
Spoke offline: nah not worth it.

**Comment by connorjclark** [comment]
Maybe we should also have a `maxTotalAttributeLength` (random pick: 600) and once we hit it we elide more/all of the value?

**Comment by Beytoven** [comment]
> Maybe we should also have a `maxTotalAttributeLength` (random pick: 600) and once we hit it we elide more/all of the value?

That would be nice, but then we run the risk of more important identifiers being cut off because they fall behind the other super long data attributes. I think the reason we opted for truncating per attribute was so that wouldn't happen. An alternative is that we could omit data-* attributes. They tend to be long and I don't know if they're actually useful in regards to element identification.

**Comment by connorjclark** [comment]
spoke offline:

* let's count the length of the attribute name + value as we go, and once it hit's ~500-600, we stop and just elide the rest of the node with `…`
* considered making sure we process `class` and `id` (most important for identification) first, but we figure that those attributes are typically already first in the markup, so we can just rely on the order of attributes correlating to importance.
* As a follow up PR, @Beytoven wants to explore making the node detail type expandable in the report, such that we could show the _entire_ HTML, unelided.

**Comment by paulirish** [comment]
> * let's count the length of the attribute name + value as we go, and once it hit's ~500-600, we stop and just elide the rest of the node with `…`

do we need to? since this adds so much more complexity i'd rather defer until we across a real example that necessitates it.

**Comment by connorjclark** [comment]
> > * let's count the length of the attribute name + value as we go, and once it hit's ~500-600, we stop and just elide the rest of the node with `…`
> 
> do we need to? since this adds so much more complexity i'd rather defer until we across a real example that necessitates it.

Can you elaborate? the example in the OP is exactly why this is necessary. Just having a cap on the size of individual attributes does not prevent a lengthy node snippet when there are many attributes.

**Comment by Beytoven** [comment]
So the approach I ended up taking here is to establish an "attribute character budget" of sorts. We simply keep count of the number of characters used for attribute names/values and once we've exceeded that budget we remove the remaining attributes. I'm seeing 50-60% reductions in snippet length for some of the longer snippets (1200+ characters).

File: lighthouse-core/lib/page-functions.js [hunk 3]
```diff
@@ -123,8 +125,25 @@ function getOuterHTMLSnippet(element, ignoreAttrs = []) {
...
+          clone.setAttribute(attributeName, attributeValue);
+        }
+        charCount += attributeName.length + attributeValue.length;
+      }
+    }
+
     const reOpeningTag = /^[\s\S]*?>/;
     const match = clone.outerHTML.match(reOpeningTag);
+    if (match && match[0] && charCount > SNIPPET_CHAR_LIMIT) {
+      return match[0].slice(0, match[0].length - 1) + '…>';
```
Comment (connorjclark) [inline lighthouse-core/lib/page-functions.js:145]: can you add some tests for this part?

Comment (Beytoven) [reply lighthouse-core/lib/page-functions.js:145]: yep, currently working on it

File: lighthouse-core/lib/page-functions.js [hunk 4]
```diff
@@ -123,8 +124,25 @@ function getOuterHTMLSnippet(element, ignoreAttrs = []) {
...
+        if (attributeValue.length > ATTRIBUTE_CHAR_LIMIT) {
+          attributeValue = attributeValue.slice(0, ATTRIBUTE_CHAR_LIMIT - 1) + '…';
+          clone.setAttribute(attributeName, attributeValue);
+        }
+        charCount += attributeName.length + attributeValue.length;
+      }
+    }
+
     const reOpeningTag = /^[\s\S]*?>/;
     const match = clone.outerHTML.match(reOpeningTag);
```
Comment (connorjclark) [inline lighthouse-core/lib/page-functions.js:142]: maybe replace with

```
const [match] = clone.outerHTML.match(reOpeningTag) || [];
```

and then drop all the `match && match[0]` for just `match` ?

File: lighthouse-core/lib/page-functions.js [hunk 5]
```diff
@@ -123,8 +124,25 @@ function getOuterHTMLSnippet(element, ignoreAttrs = []) {
     ignoreAttrs.forEach(attribute =>{
       clone.removeAttribute(attribute);
     });
+    let charCount = 0;
+    for (const attributeName of clone.getAttributeNames()) {
+      if (charCount >= snippetCharacterLimit) {
```
Comment (connorjclark) [inline lighthouse-core/lib/page-functions.js:129]:
```suggestion
      if (charCount > snippetCharacterLimit) {
```

?

**Pull request merged**